### PR TITLE
Jaxified lens equation solver

### PR DESCRIPTION
**Pull request by ahuang314**
This PR jaxifies the gradient descent portion of the lens equation solver algorithm found in lenstronomy. This jaxified solver runs faster than the helens solver by a factor of ~3-5 due to the fact that in the helens solver, triangles in the grid share vertices, so most vertices undergo ray shooting 3-6 times each, which is a flaw not present in the gradient descent algorithm. Thus, jaxtronomy will not use the helens solver. Summary of changes:

**Significant Changes**
1. The jaxified lens equation solver drops `image_position_stochastic` as it is not very practical, nor is it used often
2. The jaxified lens equation solver drops the option to use the `scipy.minimize` function to perform the gradient descent non-linearly. This is due to the fact that the function being minimized needs to be reverse-mode-differentiable in order to use `jax.scipy.minimize`, but unfortunately it is not, due to the use of for loops with a dynamic stopping point.
4. The code is less readable due to the way jax works. In the original lenstronomy implementation, there are while loops nested inside of if-statements, and jaxifying this is possible although not elegant. Additionally, the `gradient_step` function, which was originally a recursive function, has been rewritten, as recursive functions with dynamic stopping conditions are not possible within jax. Hopefully it is not too bad

**Smaller changes**
1. displaceAbs function from util has been jaxified.
2. The LensEquationSolver and EPLMajorAxis classes have been registered as custom PyTree objects, allowing the jax compiler to recognize them. JIT-compiling individual class functions involving the `self` variable is now possible for these classes, allowing for portions of algorithms to be jaxified whenever jaxifying entire pipelines/algorithms is too difficult, impossible, or sub-optimal.
3. The `initial_guess_cut` and `num_iter_max` variables in the solver have had their default values changed to False and 1000, respectively. This is because the gradient descent portion of the algorithm now runs extremely quickly, and changing these default variables can help find solutions to the lens equation that it might miss otherwise (this is a very small run-time tradeoff for accuracy and consistency, to counteract the randomness present in the `gradient_step` function)

The runtime has been improved by about a factor of 10 (also depends on search window and pixel size, I have not done rigorous runtime tests for various parameter values).

### CODE REVIEW

**Comment by ahuang314** [comment]
Not sure why test_lens_equation_solver takes 20 minutes to run on the Github but it only takes 5-10 seconds to run on my computer

**Comment by ahuang314** [comment]
4 lines are missing coverage but those lines are either not possible or very difficult to test

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+
+
+class LensEquationSolver(object):
+    """Class to solve for image positions given lens model and source position."""
+
+    def __init__(self, lensModel):
+        """This class must contain the following definitions (with same syntax as the
+        standard LensModel() class: def ray_shooting() def hessian() def magnification()
+
+        :param lensModel: instance of a class according to
```
Comment (sibirrer) [inline jaxtronomy/LensModel/Solver/lens_equation_solver.py:33]: or has this to be an instance of the jax version?

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+            x_mins = x_mins[delta_map <= min_distance * mag * 5]
+            y_mins = y_mins[delta_map <= min_distance * mag * 5]
+            if verbose:
+                print(
+                    "The number of regions that meet the plausibility criteria are %s"
+                    % len(x_mins)
+                )
+        if len(x_mins) < 1:
+            return x_mins, y_mins
+        x_mins = np.append(
```
Comment (sibirrer) [inline jaxtronomy/LensModel/Solver/lens_equation_solver.py:156]: is this supported with JAX? Might also not be needed in lenstronomy either way (adding random candidate solutions from where to do gradient decent)

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py [hunk 3]
```diff
@@ -0,0 +1,773 @@
...
+        :param kwargs_lens: lens model parameters as keyword arguments
+        :param solver: which solver to use, can be 'lenstronomy' (default), 'analytical'
+            or 'stochastic'.
+        :param kwargs: Any additional kwargs are passed to the chosen solver, see the
+            documentation of image_position_lenstronomy, image_position_analytical and
+            image_position_stochastic
+        :returns: (exact) angular position of (multiple) images ra_pos, dec_pos in units
+            of angle
+        """
+        if solver == "lenstronomy":
```
Comment (sibirrer) [inline jaxtronomy/LensModel/Solver/lens_equation_solver.py:71]: do these IF clauses work in JAX? Or do we need a different class hierarchy such that when calling the definition, no if/else statements are needed (i.e. define it on the class level instead)?

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py [hunk 4]
```diff
@@ -0,0 +1,773 @@
...
+            return self.image_position_lenstronomy(
+                sourcePos_x, sourcePos_y, kwargs_lens, **kwargs
+            )
+        if solver == "analytical":
+            return self.image_position_analytical(
+                sourcePos_x, sourcePos_y, kwargs_lens, **kwargs
+            )
+
+    def image_position_lenstronomy(
```
Comment (sibirrer) [inline jaxtronomy/LensModel/Solver/lens_equation_solver.py:80]: just curious, you don't need a jit decorator here?

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py [hunk 5]
```diff
@@ -0,0 +1,773 @@
...
+        :param kwargs_lens: lens model parameters as keyword arguments
+        :param arrival_time_sort: bool, if True, sorts image position in arrival time (first arrival photon first listed)
+        :param magnification_limit: None or float, if set will only return image positions that have an
+         abs(magnification) larger than this number
+        :param kwargs_solver: additional kwargs to be supplied to the solver. Particularly relevant are Nmeas and Nmeas_extra
+        :returns: (exact) angular position of (multiple) images ra_pos, dec_pos in units of angle
+         Note: in contrast to the other solvers, generally the (heavily demagnified) central image will also be included, so
+         setting a proper magnification_limit is more important. To get similar behaviour, a limit of 1e-1 is acceptable
+        """
+        lens_model_list = copy.deepcopy(list(self.lensModel.lens_model_list))
```
Comment (sibirrer) [inline jaxtronomy/LensModel/Solver/lens_equation_solver.py:596]: maybe this can be a self variable? I worry that deep copying for each function call might be an issue for JAX

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py [hunk 6]
```diff
@@ -0,0 +1,773 @@
...
+
+            # removing of kwargs_lens of "CONVERGENCE" profile
+            kwargs_lens_.pop(index_convergence)
+            lens_model_list.pop(index_convergence)
+        else:
+            kwargs_lens_ = kwargs_lens
+            x_, y_ = x, y
+
+        if lens_model_list not in SUPPORTED_LENS_MODELS_ANALYTICAL:
+            raise ValueError(
```
Comment (sibirrer) [inline jaxtronomy/LensModel/Solver/lens_equation_solver.py:632]: you can do raise testing with numpy.testing assert_raise() and then specify what error should be raised for the input (i.e. ValueError)

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py [hunk 7]
```diff
@@ -0,0 +1,773 @@
...
+            arrival_time_sort=arrival_time_sort,
+            initial_guess_cut=initial_guess_cut,
+            verbose=verbose,
+            x_center=x_center,
+            y_center=y_center,
+            num_random=num_random,
+            magnification_limit=magnification_limit,
+        )
+        mag_list = []
+        for i in range(len(x_mins)):
```
Comment (sibirrer) [inline jaxtronomy/LensModel/Solver/lens_equation_solver.py:702]: you actually don't need a for-loop here. Instead:
mag_list = np.abs(self.lensModel.magnification(x_mins, y_mins, kwargs_lens))
I corrected it just in lenstronomy as well

File: jaxtronomy/Util/util.py [hunk 8]
```diff
@@ -7,11 +7,10 @@
       (when applicable) and write test functions
 """
 
-import numpy as np
-import itertools
 from lenstronomy.Util.package_util import exporter
 import jax.numpy as jnp
 from jax import jit
+from lenstronomy.Util.util import make_grid, local_minima_2d, selectBest
```
Comment (sibirrer) [inline jaxtronomy/Util/util.py:13]: just checking, you want the lenstronomy and not jaxtronomy functions here

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thank you very much @ahuang314! I have some questions that may or may not be relevant, just want to check with you. Overall, looks very good and super excited how this is moving!

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 1 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:33]: Whoops, I was not careful enough about the documentation. This one is fixed now

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 2 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:156]: I left image_position_lenstronomy and candidate_solutions unjaxified, only the gradient descent portion of the algorithm has been jaxified. I have jit-decorated the computation-heavy functions so that users can still take advantage of jax even without using jax.jit on the overall solver (e.g. if you use `image_position_from_source`, you will still get speed up without using `jax.jit(image_position_from_source)` since the subfunctions have been jit-decorated). Due to the restrictions of writing in jax, I do not think it is possible to jaxify everything.

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 3 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:71]: Same as my comment [above](https://github.com/lenstronomy/JAXtronomy/pull/14#discussion_r1694451063)

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 6 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:632]: Ah OK, will implement that

File: jaxtronomy/Util/util.py (hunk 8 above)
Comment (ahuang314) [reply jaxtronomy/Util/util.py:13]: Yes, I have not jaxified those yet. And I will not jaxify make_grid, may or may not jaxify the others

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 4 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:80]: Same as my comment above

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 3 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:71]: The main struggle with jaxifying some of these if-statements is that when using jax.lax.cond, whether the condition is true or false, the resulting output array must have the same shape, which doesn't work when you're appending or removing elements of an array. For example, here's the error you would get if you were to try to jaxify the `if magnification_limit is Not None`: `true_fun and false_fun output must have same type structure, got PyTreeDef(([*, *], [*, *], [*, *])) and PyTreeDef(([*], [*], [*]))` since if magnification_limit is None, you return the whole array, but if magnification_limit is not None, you remove some elements (jnp.where has a similar issue)

Comment (sibirrer) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:71]: @ahuang314 I think we should re-design then the solver class a bit and request these options to be put into the class instance and once the class is initialized, that there are little to no ambiguities left. It's ok to deviate a bit from what lenstronomy has. In turn, we can also change a bit the structure in lenstronomy to be better compatible with the jax version

Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:71]: @sibirrer Hmm, If I'm understanding this correctly, you're suggesting that instead of having `magnification_limit` be a function argument, we instead make it a static class variable that is defined when an instance of the class is created. That way, the `if statement` is not needed and we can go through with downselecting the solutions.

This will still run into the same fundamental problem: JAX requires that all array shapes and sizes are known at compile time. This means that functions that branch off of conditions must have the same return structures, and repeated executions of the same function must have the same return structure regardless of the function arguments. The problem with jaxifying the lens equation solver completely is that you don't know how many solutions there will be until after you've solved it, and you also don't know how many solutions are downselected from the magnification limit until you do it. This is why I did not jaxify the `image_position_lenstronomy` and `candidate_solutions` functions in the solver. Here is a screenshot showing the error you would get if you tried the downselecting:
image
image

The way it is done in herculens is, the user specifies the number of solutions `nsolutions` that the solver expects to find. This is treated as a static variable so the solver will always return exactly an array of size `nsolutions`. However, I dislike this method of doing it because if the solver actually finds more solutions than that, some will be excluded, and if the solver finds less than that, one of the solutions will be duplicated until the array is of size `nsolutions`. The magnification limit downselection in herculens is also done in an unjaxified way.

To conclude, I think we should only focus on jaxifying only the heavy-computation subfunctions and not worry about large scale restructuring in an attempt to jaxify everything. The following three line profiles show that, even if we were to fully jaxify the entirety of `image_position_lenstronomy` and `candidate_solutions`, the benefit we would get is extremely tiny due to how little time is actually spent on the currently unjaxified part of the code.
[Line profile for unjaxified solver with jaxified ray shooting](https://pastebin.com/MPShhPen)
[Line profile for solver after jaxifying gradient descent with jaxified ray shooting](https://pastebin.com/ZRRVDGmw)
[Line profile for `candidate_solutions`](https://pastebin.com/fQbeRfTF)

Comment (sibirrer) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:71]: I see. It's very tricky to jaxify all the parts even when class structure change, as the issues are algorithmically.
The only advantage of jaxifying everything would be an auto-differentiation. But since we barely use the lens equation solver in the actual fitting, we do not have to solve a problem we do not have for now.

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 7 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:702]: Done

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 5 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:596]: Currently this function is not jaxified since it the analytical solver uses the numba compiler. We can change this whenever we jaxify the analytical solver

**Review by sibirrer — APPROVED** [review]

File: jaxtronomy/LensModel/Solver/lens_equation_solver.py (hunk 5 above)
Comment (ahuang314) [reply jaxtronomy/LensModel/Solver/lens_equation_solver.py:596]: Just tested it and deepcopy does not seem to cause any issues with JAX
